keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,494 @@
+
+#### Story 3 - Administrator uses Elasticsearch to debug problem with workload
+ 
+TODO(serathius): Show examples of some queries to Elasticsearch
+
+### Performance
+
+We are expecting improving performance of logger library due ability to change logging library to more efficient, then current `klog`. 
+
+
+TODO(serathius): provide performance comparison and estimate log size change
+
+As part of this proposal we expect logging to be more dependent on usage of context, which can create overhead when adding to and passing it through call stack. We accept this overhead as cost of improving observability in K8s. There is a proposal that would benefit [Leveraging Distributed Tracing to Understand Kubernetes Object Lifecycles](https://github.com/kubernetes/enhancements/pull/650#issuecomment-552577896).
+
+### Risks and Mitigations
+
+#### Migration failing halfway
+
+Kubernetes is a huge project spread across multiple repositories with tens of thousands of logging calls. As not all migration steps can be performed automatically there is a need for some manual work to be performed to verify the results of automatic migration to new logging API and adjusting its results based on a logging migration guidelines and best practices.
+
+As all of this manual work needs to be performed before we can declare GA for structured logging there is a risk that we will not find volunteers who would like to perform this work.
+As a possible mitigation we could consider performing more migration steps automatically:
+* Implement more intelligent processing of log messages.
+* Implement automatic detection of log context by analyzing variables which are visible in the scope of logger call.
+* Implement automatic detection of common types of fields used in logs.  
+
+#### Hard to customize logger format
+
+Proposed logger implementation are not designed for extensibility. Any change of behavior would require a change in code by creating PR. Looking at kubernetes history we know that this will not be maintainable. Eventually kubernetes adapted dedicated interfaces like CSI, CRI, CNI. We think that similar think could happen for logging in the future, but it doesn’t need to be introduced immediately. A good point for starting this discussion would be when this effort would be GAed. Then we could propose another logging API implementation.
+
+#### Huge increase of log volume
+
+This effort should result in reducing costs of log storage and analysis in production setups. Introducing structured logs will allow for creating better indices and reducing the size of data needed for analysis. From log throughput perspective there will be an increase of log size causing a pressure of log ingestion (disk, logging agents, logging API). To reduce the potential impact we will consider adding additional configuration flags allowing user to specify which metadata should be dropped
```

Comment:
I would like to see some data/examples here.

Reply:
I think there will be a slight increase. Messages and structured fields in a log line tend to be somewhat duplicate in wording.